src/zal.rs

```diff
@@ -0,0 +1,185 @@
+//! This module provides "ZK Acceleration Layer" traits
+//! to abstract away the execution engine for performance-critical primitives.
+//!
+//! The ZAL Engine is voluntarily left unconstrained
+//! so that accelerator libraries are not prematurely limited.
+//!
+//! Terminology
+//! -----------
+//!
+//! We use the name Backend+Engine for concrete implementations of ZalEngine.
+//! For exaple H2cEngine for pure Halo2curves implementation.
+//!
+//! Alternative names considered were Executor or Driver however
+//! - executor is already used in Rust (and the name is long)
+//! - driver will be confusing as we work quite low-level with GPUs and FPGAs.
+//!
+//! Unfortunately Engine is used in bn256 for pairings.
+//! Fortunately ZalEngine is only used in the prover
+//! while "pairing engine" is only used in the verifier
+//!
+//! Initialization design space
+//! ---------------------------
+//!
+//! It is recommended that ZAL backends provide:
+//! - an initialization function:
+//!   - either "fn new() -> ZalEngine" for simple libraries
+//!   - or a builder pattern for complex initializations
+//! - a shutdown function.
+//!
+//! The ZalEngine can be a stub type
+//! and the shutdown function might be unnecessary
+//! if the ZalEngine uses a global threadpool like Rayon.
+//!
+//! Backends might want to add as an option:
+//! - The number of threads (CPU)
+//! - The device(s) to run on (multi-sockets machines, multi-GPUs machines, ...)
+//! - The curve (JIT-compiled backend)
+
+use crate::msm::best_multiexp;
+use pasta_curves::arithmetic::CurveAffine;
+
+// The ZK Accel Layer API
+// ---------------------------------------------------
+
+pub trait ZalEngine {}
+
+pub trait MsmAccel<C: CurveAffine>: ZalEngine {
+    fn msm(&self, coeffs: &[C::Scalar], base: &[C]) -> C::Curve;
+
+    // Caching API
+    // -------------------------------------------------
+    // From here we propose an extended API
+    // that allows reusing coeffs and/or the base points
+    //
+    // This is inspired by CuDNN API (Nvidia GPU)
+    // and oneDNN API (CPU, OpenCL) https://docs.nvidia.com/deeplearning/cudnn/api/index.html#cudnn-ops-infer-so-opaque
+    // usage of descriptors
+    //
+    // https://github.com/oneapi-src/oneDNN/blob/master/doc/programming_model/basic_concepts.md
+    //
+    // Descriptors are opaque pointers that hold the input in a format suitable for the accelerator engine.
+    // They may be:
+    // - Input moved on accelerator device (only once for repeated calls)
+    // - Endianess conversion
+    // - Converting from Montgomery to Canonical form
+    // - Input changed from Projective to Jacobian coordinates or even to a Twisted Edwards curve.
+    // - other form of expensive preprocessing
+    type CoeffsDescriptor<'c>;
+    type BaseDescriptor<'b>;
+
+    fn get_coeffs_descriptor<'c>(&self, coeffs: &'c [C::Scalar]) -> Self::CoeffsDescriptor<'c>;
```

Comment:
Another thing that comes to my mind related to the descriptors, shouldn't there be also methods to free the resources allocated for the descriptors?

Reply:
In general the descriptors should be plain old data, I don't see a use-case for non-`Copy` descriptors

see my comment in the document:

            The most complex descriptors should be:
            - Pointer(s) to data
            - size of data
            - data in layout (for example canonical or bit-reversed permuted)
            - data out layout
            - enum or integer ID for operation (MSM, FFT, Coset FFT)
            - Device(s) ID